### PR TITLE
CHAT-838 Possibility of Leaving Rooms in Chat

### PR DESCRIPTION
**Pull request by nttuyen**
- Add button to allow user to leave the team chat room . 
- Spec: https://community.exoplatform.com/portal/g/:spaces:platform_4/exo_platform/wiki/Leave_Chat_Room

### CODE REVIEW

File: application/src/main/resources/locale/portlet/chat/Resource.properties [hunk 1]
```diff
@@ -163,6 +164,7 @@ exoplatform.chat.timeout.login=Login
 exoplatform.chat.team.help=Ex: "ro" or "Ro Br" to search Robert Bruce
 exoplatform.chat.team.msg.adduser={0} added {1} to this conversation.
 exoplatform.chat.team.msg.removeuser={0} removed {1} from this room.
+exoplatform.chat.team.msg.leaveteam={0} has leaved this room.
```
Comment (thomasdelhomenie) [inline application/src/main/resources/locale/portlet/chat/Resource.properties]: "leaved" -> "left"

File: application/src/main/resources/locale/portlet/chat/Resource_en.properties [hunk 2]
```diff
@@ -163,6 +164,7 @@ exoplatform.chat.timeout.login=Login
 exoplatform.chat.team.help=Ex: "ro" or "Ro Br" to search Robert Bruce
 exoplatform.chat.team.msg.adduser={0} added {1} to this conversation.
 exoplatform.chat.team.msg.removeuser={0} removed {1} from this room.
+exoplatform.chat.team.msg.leaveteam={0} has leaved this room.
```
Comment (thomasdelhomenie) [inline application/src/main/resources/locale/portlet/chat/Resource_en.properties]: "leaved" -> "left"

File: application/src/main/webapp/js/chatRoom.js [hunk 3]
```diff
@@ -1034,6 +1034,9 @@
       } else if (options.type==="type-remove-team-user") {
         var users = "<b>" + options.users.replace("; ","</b>; <b>") + "</b>";
         out += chatBundleData["exoplatform.chat.team.msg.removeuser"].replace("{0}", "<b>" + options.fullname + "</b>").replace("{1}", users);
+      } else if (options.type==="type-leave-team-user") {
+        var users = "<b>" + options.users.replace("; ","</b>; <b>") + "</b>";
+        out += chatBundleData["exoplatform.chat.team.msg.leaveteam"].replace("{0}", "<b>" + options.fullname + "</b>");
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chatRoom.js]: I would use "exoplatform.chat.team.msg.leaveroom" instead.

File: application/src/main/webapp/js/chat.js [hunk 4]
```diff
@@ -446,6 +447,41 @@
...
+
+        success: function (response) {
+          if (typeof callback === "function") {
+            callback(response);
+          }
+        },
+
+        error: function (xhr, status, error) {
+          chatApplication.renderRooms();
+          alert(xhr.responseText);
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js]: To be removed.

File: application/src/main/webapp/js/chat.js [hunk 5]
```diff
@@ -446,6 +447,41 @@
...
+        },
+
+        success: function (response) {
+          if (typeof callback === "function") {
+            callback(response);
+          }
+        },
+
+        error: function (xhr, status, error) {
+          chatApplication.renderRooms();
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js]: Why do we need to re-render rooms in this case ?

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java [hunk 6]
```diff
@@ -743,6 +676,126 @@ public ChatServer()
...
+  @Route("/removeRoomUser")
+  public Response removeRoomUser(String user, String token, String roomId, String dbName) {
+    if (!tokenService.hasUserWithToken(user, token, dbName)) {
+      return Response.notFound("Please login !");
+    }
+
+    if (user != null && !user.isEmpty()) {
+      RoomBean roomBean = chatService.getTeamRoomById(roomId, dbName);
+      if (roomBean == null) {
+        LOG.log(Level.SEVERE, "Can't remove user " + user + "from Room id" + roomId + ". Room not found!");
```
Comment (thomasdelhomenie) [inline server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: Missing spaces "from Room id" -> " from Room id "
Check others logs as well.

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java [hunk 7]
```diff
@@ -743,6 +676,126 @@ public ChatServer()
...
+    if (newRoomUsers != null && !newRoomUsers.isEmpty()) {
+      List<String> usersToAdd = new JSONArray();
+      usersToAdd.addAll(newRoomUsers);
+      List<String> usersToRemove = new JSONArray();
+
+      JSONObject data = new JSONObject();
+      data.put("title", teamName);
+      data.put("members", usersToAdd);
+      RealTimeMessageBean updatedRoomMessage = new RealTimeMessageBean(
+              RealTimeMessageBean.EventType.ROOM_UPDATED,
```
Comment (thomasdelhomenie) [inline server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: Shouldn't we use a different event type when an user leaves a room ? It is not the same action, and therefore the same message, so I think it should be a dedicated event type. WDYT ?

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java [hunk 8]
```diff
@@ -743,6 +676,126 @@ public ChatServer()
                    .withCharset(Tools.UTF_8);
   }
 
+  @Resource
+  @Route("/removeRoomUser")
+  public Response removeRoomUser(String user, String token, String roomId, String dbName) {
```
Comment (thomasdelhomenie) [inline server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: Can you add unit tests for this new method please ?

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java (hunk 7 above)
Comment (nttuyen) [reply server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: is that ok if I update the event to ROOM_MEMBER_LEFT("room-member-left"), we already have this event type

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java (hunk 8 above)
Comment (nttuyen) [reply server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: yes, let me add

File: server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java (hunk 7 above)
Comment (thomasdelhomenie) [reply server-embedded/src/main/java/org/exoplatform/chat/server/ChatServer.java]: Yes, ROOM_MEMBER_LEFT is the right event to use :+1:

**Comment by hghorbel** [comment]
> the icon of the Leave Room action is the same as the Edit Room action. We should have a different icon 
We don't have any icon to display, we just have a label in the dropdown. 

> there is also the rule LEAVE_06 "When the Room manager invite once again a user that left the room before, this user will see only new comments sent, like he is a new Room member. " but that's actually not the current behavior (when I am invited in a room I can see all old messages), so I don't think we should implement it. WDYT ?
We can use the current behavior as first feature draft and investigate about it with customers.

**Comment by thomasdelhomenie** [comment]
> We don't have any icon to display, we just have a label in the dropdown.

Not in the current version. You talk about something that is not implemented yet (Chat redesign). We need an icon please.

> We can use the current behavior as first feature draft and investigate about it with customers.

Ok, please update the spec.

**Comment by thomasdelhomenie** [comment]
Thanks @hghorbel !

**Comment by nttuyen** [comment]
@hghorbel , @thomasdelhomenie , i'm done fixing all issues following your comments, pls check

File: application/src/main/resources/locale/portlet/chat/Resource.properties [hunk 9]
```diff
@@ -163,6 +164,11 @@ exoplatform.chat.timeout.login=Login
 exoplatform.chat.team.help=Ex: "ro" or "Ro Br" to search Robert Bruce
 exoplatform.chat.team.msg.adduser={0} added {1} to this conversation.
 exoplatform.chat.team.msg.removeuser={0} removed {1} from this room.
+exoplatform.chat.team.leave.title=Leave Room
+exoplatform.chat.team.leave.ok=Yes
+exoplatform.chat.team.leave.ko=No
+exoplatform.chat.team.leave.message=You are about to leave the "{0}" Room. You will loose access to all messages it contains. Are you sure you want to leave ?
```
Comment (thomasdelhomenie) [inline application/src/main/resources/locale/portlet/chat/Resource.properties]: I think we should simplify the message to : "Are you sure you want to leave the room "{0}" ?"

Comment (nttuyen) [reply application/src/main/resources/locale/portlet/chat/Resource.properties]: @thomasdelhomenie , i'm done updating message translation

**Pull request merged**